src/Cardano/Wallet/CoinSelection.hs

```diff
@@ -0,0 +1,154 @@
+    }
+
+
+data CoinSelHardErr =
+    CoinSelHardErrOutputCannotCoverFee Text Text
+    -- ^ Payment to receiver insufficient to cover fee
+    -- We record the original output and the fee it needed to cover.
+    -- Only applicable using 'ReceiverPaysFees' regulation
+  | CoinSelHardErrOutputIsRedeemAddress Text
+    -- ^ Attempt to pay into a redeem-only address
+  | CoinSelHardErrCannotCoverFee
+    -- ^ UTxO exhausted whilst trying to pick inputs to cover remaining fee
+  | CoinSelHardErrMaxInputsReached Text
+    -- ^ When trying to construct a transaction, the max number of allowed
+    -- inputs was reached.
+  | CoinSelHardErrUtxoExhausted Text Text
+    -- ^ UTxO exhausted during input selection
+    -- We record the balance of the UTxO as well as the size of the payment
+    -- we tried to make.
+    -- See also 'CoinSelHardErrCannotCoverFee'
+
+type CoinSelPolicy utxo m a =
+       NonEmpty TxOut
+    -> utxo
+    -> m (Either CoinSelHardErr a)
+
+
+newtype MaximumNumOfInputs =
+    MaximumNumOfInputs { getNum :: Word64 }
+
+data CoinSelFinalResult = CoinSelFinalResult
+    { csrInputs  :: NonEmpty (TxIn, TxOut)
+    -- ^ Picked inputs
+    , csrOutputs :: NonEmpty TxOut
+    -- ^ Picked outputs
+    , csrChange  :: [Coin]
+    -- ^ Resulting changes
+    }
+
+
+-- | Random input selection policy
+random
+    :: forall m. MonadRandom m
+    => CoinSelectionOptions
+    -> MaximumNumOfInputs
+    -> CoinSelPolicy UTxO m CoinSelFinalResult
+random _ = undefined
+
+-- | Largest-first input selection policy
+largestFirst
+    :: forall m. MonadRandom m
+    => CoinSelectionOptions
+    -> MaximumNumOfInputs
+    -> CoinSelPolicy UTxO m CoinSelFinalResult
+largestFirst _ = undefined
+
+
+----------------------------------------------------------------------------
+--                        Fee related                                     --
+----------------------------------------------------------------------------
+
+newtype Fee = Fee { getFee :: Coin }
+
+estimateCardanoFee
+    :: TxSizeLinear
+    -- ^ linear fee policy
+    -> Int
+    -- ^ number of inputs
+    -> [Word64]
+    -- ^ coin value of each output
+    -> Word64
+estimateCardanoFee _ = undefined
+
+estimateMaxTxInputs
+  :: ByteString
+    -- ^ Maximum size of a transaction
+  -> Word64
+estimateMaxTxInputs _ = undefined
+
+----------------------------------------------------------------------------
+--                        Helpers                                         --
+----------------------------------------------------------------------------
+
+-- | A fractional coefficient of fixed precision.
+newtype Coeff = Coeff Nano
+    deriving (Eq, Ord, Show, Generic, NFData, Num)
+
+-- | A linear equation on the transaction size. Represents the @\s -> a + b*s@
+-- function where @s@ is the transaction size in bytes, @a@ and @b@ are
+-- constant coefficients.
+data TxSizeLinear = TxSizeLinear !Coeff !Coeff
+    deriving (Eq, Ord, Show, Generic)
```

Comment:
Same as above, not needed for the coin selection so-to-speak, this is about fee estimation.

Reply:
removed